### PR TITLE
Feat/cluster server

### PR DESCRIPTION
**Pull request by iGxnon**
Reopen of #434 
Base on #476 

Due to the misoperation of forced push during rebase, the commit record was truncated and automatically closed by Github. I reopen here.

Please briefly answer these questions:

* what problem are you trying to solve? (or if there's no problem, what's the motivation for this change?)

* what changes does this pull request make?

* are there any non-obvious implications of these changes? (does it break compatibility with previous versions, etc)

### CODE REVIEW

File: xline/src/server/maintenance.rs [hunk 1]
```diff
@@ -34,17 +37,28 @@ where
...
+    client: Arc<Client<Command>>,
+    /// cluster information
+    cluster_info: Arc<ClusterInfo>,
 }
 
 impl<S> MaintenanceServer<S>
 where
     S: StorageApi,
 {
     /// New `LeaseServer`
```
Comment (Phoenix500526) [inline xline/src/server/maintenance.rs:50]: Please correct this comment.

File: xline/src/server/maintenance.rs [hunk 2]
```diff
@@ -67,9 +81,31 @@ where
...
+            .members
+            .iter()
+            .find(|member| member.id == self_id)
+            .map_or(false, |member| member.is_learner);
+        let response = StatusResponse {
+            header: Some(header),
+            version: "NA".to_owned(),
+            db_size: -1,
+            leader: cluster.leader_id.unwrap_or(0), // None means this member believes there is no leader
+            raft_index: 0,                          // TODO
```
Comment (Phoenix500526) [inline xline/src/server/maintenance.rs:101]: Todo what?

**Review by Phoenix500526 — COMMENTED** [review]
This PR should be based on #437. Please add this information to the pr description.

**Comment by iGxnon** [comment]
> mation to the pr description.

This PR should be based on the last PR for the entire membership feature. I changed its base from the previous one to the current one that can be reviewed to avoid too many draft PRs piling up.

File: xline/src/server/maintenance.rs (hunk 1 above)
Comment (iGxnon) [reply xline/src/server/maintenance.rs]: done

File: xline/src/server/maintenance.rs (hunk 2 above)
Comment (iGxnon) [reply xline/src/server/maintenance.rs]: done

File: xline/src/server/maintenance.rs [hunk 3]
```diff
@@ -67,9 +81,32 @@ where
...
+        let self_id = self.cluster_info.self_id();
+        let is_learner = cluster
+            .members
+            .iter()
+            .find(|member| member.id == self_id)
+            .map_or(false, |member| member.is_learner);
+        // TODO should we add `raft_index` and `raft_applied_index` in FetchClusterResponse?
+        let response = StatusResponse {
+            header: Some(header),
+            version: "NA".to_owned(),
```
Comment (bsbds) [inline xline/src/server/maintenance.rs:99]: `version` should be Xline version. Please refer to etcd and fill other fields of this repsonse.

File: xline/src/server/maintenance.rs [hunk 4]
```diff
@@ -67,9 +81,32 @@ where
...
         _request: tonic::Request<StatusRequest>,
     ) -> Result<tonic::Response<StatusResponse>, tonic::Status> {
-        Err(tonic::Status::unimplemented(
-            "status is unimplemented".to_owned(),
-        ))
+        let cluster = self
+            .client
+            .get_cluster_from_curp(false)
+            .await
+            .map_err(propose_err_to_status)?;
```
Comment (themanforfree) [inline xline/src/server/maintenance.rs:88]: Status should be get directly from the local server, not through client.

File: xline/src/server/maintenance.rs [hunk 5]
```diff
@@ -67,9 +81,32 @@ where
...
+        let header = self.header_gen.gen_header();
+        let self_id = self.cluster_info.self_id();
+        let is_learner = cluster
+            .members
+            .iter()
+            .find(|member| member.id == self_id)
+            .map_or(false, |member| member.is_learner);
+        // TODO should we add `raft_index` and `raft_applied_index` in FetchClusterResponse?
```
Comment (themanforfree) [inline xline/src/server/maintenance.rs:96]: Same as above, Could we add a set of interfaces to the curp server to get state externally

File: xline/src/server/cluster_server.rs [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+};
+
+use super::command::{propose_err_to_status, Command};
+use crate::header_gen::HeaderGenerator;
+use crate::storage::storage_api::StorageApi;
+
+/// Cluster Server
+pub(crate) struct ClusterServer<S>
+where
+    S: StorageApi,
```
[the remark this answers is not in the document]
Comment (iGxnon) [reply xline/src/server/cluster_server.rs:22]: The same generic type also appears in `lock_server` and `auth_server`, why to keep that? @themanforfree

File: xline/src/server/maintenance.rs (hunk 4 above)
Comment (iGxnon) [reply xline/src/server/maintenance.rs:88]: Yes, the `false` parameter in `get_cluster_from_curp` will trigger to fetch status from local server first, if failed, then fetch the leader status.

File: xline/src/server/cluster_server.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (themanforfree) [reply xline/src/server/cluster_server.rs:22]: `command_from_request_wrapper` need the generic type `S`, `auth_server` and `lock_server` need to use this function

File: xline/src/server/maintenance.rs (hunk 5 above)
Comment (iGxnon) [reply xline/src/server/maintenance.rs:96]: I think this interface is similar to `fetch cluster`. We can add more fields to the fetch cluster response.

Comment (iGxnon) [reply xline/src/server/maintenance.rs]: After bypassing the RPC sent to oneself within the curp client, more detailed information can be obtained using the curp node inside the client. I will handle this in the upcoming PR.

File: xline/src/server/maintenance.rs (hunk 4 above)
Comment (iGxnon) [reply xline/src/server/maintenance.rs]: see above

File: xline/src/server/maintenance.rs (hunk 3 above)
Comment (iGxnon) [reply xline/src/server/maintenance.rs]: done

File: xline/src/server/cluster_server.rs (hunk 6 above)
[the remark this answers is not in the document]
Comment (iGxnon) [reply xline/src/server/cluster_server.rs]: done

**Review by bsbds — APPROVED** [review]

**Review by themanforfree — APPROVED** [review]
LGTM

**Pull request merged**